### PR TITLE
cluster-autoscaler: Dynamic Reconfiguration via ConfigMaps

### PR DESCRIPTION
**Pull request by mumoshu**
This is the first iteration to address #2181.
Please read the discussion in the issue to see more context.

@andrewsykim @Raffo @mwielgus Would you mind reviewing this? :bow:

TODOs:

- [x] Add unit tests

Non-TODOs:

- Automatic creation of configmaps filled with defaults, like cluster-proportional-autoscaler does with `--configmap` with `--default-params`. IMHO doing so is a completely orthogonal issue therefore won't be addressed in this PR.

---

This PR adds a new optional flag named `configmap` to specify the name of a configmap containing node group specs.

The configmap is polled every `scan-interval` seconds to reconfigure cluster-autoscaler dynamically at runtime.

Example usage:

```
./cluster-autoscaler --v=4 --cloud-provider=aws --skip-nodes-with-local-storage=false --logtostderr --leader-elect=false --configmap=cluster-autoscaler --logtostderr
```

The configmap would look like:

```yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: cluster-autoscaler
  namespace: kube-system
data:
  settings: |-
    {
      "nodeGroups": [
        {
          "minSize": 1,
          "maxSize": 2,
          "name": "kubeawstest-nodepool1-AutoScaleWorker-1VWD4GAVG35L5"
        }
      ]
    }
 ```

### CODE REVIEW

File: cluster-autoscaler/cluster_autoscaler.go [hunk 1]
```diff
@@ -73,13 +72,16 @@ const (
 	MostPodsExpanderName = "most-pods"
 	// LeastWasteExpanderName selects a node group that leaves the least fraction of CPU and Memory
 	LeastWasteExpanderName = "least-waste"
+	// Namespace in which cluster-autoscaler run
+	KubernetesNamespace = "kube-system"
```
Comment (andrewsykim) [inline cluster-autoscaler/cluster_autoscaler.go]: Running cluster autoscaler in kube-system namespace is recommended but not required.

File: cluster-autoscaler/cluster_autoscaler.go [hunk 2]
```diff
@@ -410,7 +380,7 @@ func main() {
 		kube_leaderelection.RunOrDie(kube_leaderelection.LeaderElectionConfig{
 			Lock: &resourcelock.EndpointsLock{
 				EndpointsMeta: apiv1.ObjectMeta{
-					Namespace: "kube-system",
```
Comment (andrewsykim) [inline cluster-autoscaler/cluster_autoscaler.go]: I guess this ~is~ was a bug for assuming kube-system namespace :P

File: cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
+package builder
+
+import (
+	"github.com/golang/glog"
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider/aws"
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider/gce"
+	"os"
```
Comment (andrewsykim) [inline cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go]: standard lib imports go first.

File: cluster-autoscaler/cluster_autoscaler.go [hunk 4]
```diff
@@ -45,6 +42,8 @@ import (
 	"github.com/golang/glog"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/spf13/pflag"
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider/builder"
```
Comment (andrewsykim) [inline cluster-autoscaler/cluster_autoscaler.go]: any reason why these imports are separated from the other `k8s` imports?

File: cluster-autoscaler/dynamic/config.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
+package dynamic
```
Comment (andrewsykim) [inline cluster-autoscaler/dynamic/config.go]: there's already a `cluster-autoscaler/config` package, wondering if this should all go in there, or if `dynamic` should be a subpackage of config.

File: cluster-autoscaler/dynamic_reconfiguration.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	return DynamicReconfiguration{
+		configMapName:        configMapName,
+		autoscalingContext:   autoscalingContext,
+		configFetcher:        configFetcher,
+		cloudProviderBuilder: cloudProviderBuilder,
+	}
+}
+
+// Run dynamic reconfiguration once
+func (r DynamicReconfiguration) Run() error {
```
Comment (andrewsykim) [inline cluster-autoscaler/dynamic_reconfiguration.go]: should the receiver be `r *DynamicReconfiguration` since we're updating its autoscalingContext?

File: cluster-autoscaler/dynamic/config.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+// Config which represents not static but dynamic configuration of cluster-autoscaler which would be updated periodically at runtime
+type Config struct {
+	Settings
+	resourceVersion string
+}
+
+// Settings of cluster-autoscaler contained in the latest config, which should be consumed by cluster-autoscaler
+type Settings struct {
```
Comment (andrewsykim) [inline cluster-autoscaler/dynamic/config.go]: What's the reason for having a new struct for storing `NodeGroups`? Can't it just be another field in `Config` struct?

File: cluster-autoscaler/dynamic/config_fetcher.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+		configMapName: configMapName,
+		namespace:     namespace,
+		kubeClient:    kubeClient,
+		lastConfig:    NewDefaultConfig(),
+	}
+}
+
+// Returns the config if it has changed since the last sync. Returns nil if it has not changed.
+func (c *ConfigFetcher) FetchConfigIfUpdated() (*Config, error) {
+	cm, err := c.kubeClient.Core().ConfigMaps(c.namespace).Get(c.configMapName)
```
Comment (andrewsykim) [inline cluster-autoscaler/dynamic/config_fetcher.go]: If --configmap is set but the config map doesn't exist should we just create one?

**Review by andrewsykim — COMMENTED** [review]
I left a few addressable comments and some questions. Didn't leave too many comments regarding dynamic config logic since I'm still wrapping my head around that. I'll try get another review in later this week.

File: cluster-autoscaler/dynamic/config.go (hunk 5 above)
Comment (andrewsykim) [inline cluster-autoscaler/dynamic/config.go]: After reading the contents of that package it doesn't seem like its being used cluster autoscaler config anyways.. maybe it should but might be outside the scope of this PR. I'll let the maintainers decide :P

File: cluster-autoscaler/dynamic/config_fetcher.go (hunk 8 above)
Comment (mumoshu) [inline cluster-autoscaler/dynamic/config_fetcher.go]: That's listed in my non-TODOs https://github.com/kubernetes/contrib/pull/2226#issue-197912602 for now.

If you'd like to do so, I'm ok with creating one with empty `nodeGroups` though.

File: cluster-autoscaler/dynamic_reconfiguration.go (hunk 6 above)
Comment (mumoshu) [inline cluster-autoscaler/dynamic_reconfiguration.go]: Probably no? because what's we're updating is not `DynamicConfiguration` but rather `AutoscalingContext`.

File: cluster-autoscaler/dynamic/config.go (hunk 7 above)
Comment (mumoshu) [inline cluster-autoscaler/dynamic/config.go]: That's because I'd wanted to differentiate settings and their version. Settings are loaded from  a JSON object associated to the `settings` key in a configmap while a version is loaded from the metadata of it.

File: cluster-autoscaler/cluster_autoscaler.go (hunk 2 above)
Comment (mumoshu) [inline cluster-autoscaler/cluster_autoscaler.go]: Ah, understood! Would you like me to fix the bug in this PR?

File: cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go (hunk 3 above)
Comment (mumoshu) [inline cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go]: Thanks

File: cluster-autoscaler/cluster_autoscaler.go (hunk 4 above)
Comment (mumoshu) [inline cluster-autoscaler/cluster_autoscaler.go]: No. I'll fix it

File: cluster-autoscaler/dynamic/config_fetcher.go (hunk 8 above)
Comment (andrewsykim) [inline cluster-autoscaler/dynamic/config_fetcher.go]: Sorry, I totally missed that. I'm not exactly sure yet what the desired behaviour should be. I'll think about it and get back to you.

File: cluster-autoscaler/cluster_autoscaler.go [hunk 9]
```diff
@@ -73,13 +72,16 @@ const (
...
+	// Namespace in which cluster-autoscaler run
+	KubernetesNamespace = "kube-system"
 )
 
 var (
 	nodeGroupsFlag          MultiStringFlag
 	address                 = flag.String("address", ":8085", "The address to expose prometheus metrics.")
 	kubernetes              = flag.String("kubernetes", "", "Kuberentes master location. Leave blank for default")
 	cloudConfig             = flag.String("cloud-config", "", "The path to the cloud provider configuration file.  Empty string for no configuration file.")
+	configMapName           = flag.String("configmap", "", "The name of the ConfigMap containing settings used for dynamic reconfiguration. Empty string for no ConfigMap.")
```
Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: Either provide namespace information via downward api or include it here.

File: cluster-autoscaler/nodegroup/spec.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
+package nodegroup
```
Comment (mwielgus) [inline cluster-autoscaler/nodegroup/spec.go]: please keep it in the config/dynamic package. No reason for creating another top-level package.

File: cluster-autoscaler/nodegroup/spec.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+	return &spec, nil
+}
+
+// Produces an error if there's an invalid field in the node group spec
+func (s Spec) Validate() error {
+	if s.MinSize <= 0 {
+		return fmt.Errorf("min size must be >= 1")
+	}
+
```
Comment (mwielgus) [inline cluster-autoscaler/nodegroup/spec.go]: remove empty lines.

File: cluster-autoscaler/nodegroup/spec.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
+package nodegroup
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// A node group spec
+type Spec struct {
```
Comment (mwielgus) [inline cluster-autoscaler/nodegroup/spec.go]: NodeGroupSpec or something similar.

File: cluster-autoscaler/dynamic/config_fetcher.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch config map named %s: %v", c.configMapName, err)
+	}
+
+	configFromServer, err := ConfigFromConfigMap(cm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load dyamic config: %v", err)
```
Comment (mwielgus) [inline cluster-autoscaler/dynamic/config_fetcher.go]: Add an event for the config map object.

File: cluster-autoscaler/dynamic/config_fetcher.go (hunk 8 above)
Comment (mwielgus) [inline cluster-autoscaler/dynamic/config_fetcher.go]: I would rather raise a loud error. There are couple reasons why the config map is not there, like
* User misspelled the name/namespace
* User has not added it yet.
But in none of the cases i can think of adding an empty map actually helps the user.

File: cluster-autoscaler/cluster_autoscaler.go (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: Yeah, lets fix it. Provide a flag with namespace and populate it via downward api. The flag is useful in this PR anyway.

File: cluster-autoscaler/cluster_autoscaler.go [hunk 14]
```diff
@@ -222,13 +180,25 @@ func run(_ <-chan struct{}) {
...
 				updateLastTime("main")
 
+				if dynamicReconfiguration.Enabled() {
+					reconfigureStart := time.Now()
+					updateLastTime("reconfigure")
+					if err := dynamicReconfiguration.Run(); err != nil {
+						glog.Errorf("Failed to reconfigure : %v", err)
+					}
+					updateDuration("reconfigure", reconfigureStart)
+				}
```
Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: For safety, any config change should stop and recreate all the stuff running in CA. I imagine that we would like to change not only the node groups but also thresholds and other settings.

File: cluster-autoscaler/dynamic/config.go [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/contrib/cluster-autoscaler/nodegroup"
+	"k8s.io/kubernetes/pkg/api/v1"
+)
+
+// Config which represents not static but dynamic configuration of cluster-autoscaler which would be updated periodically at runtime
+type Config struct {
+	Settings
+	resourceVersion string
```
Comment (mwielgus) [inline cluster-autoscaler/dynamic/config.go]: Why resource version is not public?

File: cluster-autoscaler/dynamic/config.go [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+	NodeGroups []nodegroup.Spec `json:"nodeGroups"`
+}
+
+// NewDefaultConfig builds a new config object
+func NewDefaultConfig() Config {
+	return Config{
+		Settings: Settings{
+			NodeGroups: []nodegroup.Spec{},
+		},
+		resourceVersion: "",
```
Comment (mwielgus) [inline cluster-autoscaler/dynamic/config.go]: ResourceVersion = 1?

**Comment by mwielgus** [comment]
I would prefer if the configuration change triggered full recreate of the CA structures. The current approach works with node groups but will cause hard to debug/understand  problems if we try to add thresholds, constants, etc to the config map.

File: cluster-autoscaler/cluster_autoscaler.go (hunk 2 above)
Comment (andrewsykim) [inline cluster-autoscaler/cluster_autoscaler.go]: We'll have to remember to update cluster autoscaler examples/manifests once we upgrade to a version that expects the namespace flag.

Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: The flag will have a default value of kube-system, but yes, we will have to update exmples/k8s manifests with the next release.

**Comment by mumoshu** [comment]
@mwielgus @andrewsykim Thanks for your reviews!
I've addressed all the comments. I'll soon try to rebase this to the latest master and start testing in real kubernetes clusters but could you mind taking a look before I go too deep?

@mwielgus To achieve what you've requested in your last comment https://github.com/kubernetes/contrib/pull/2226#issuecomment-269617410, I had to make a big structural change to the code base https://github.com/kubernetes/contrib/pull/2226/commits/3f9e4ae55325537bf100c480a4e0f4794f69c0ad

Several points to be noted:

* Introduced `autoscaler/` and moved  all the `dynamic` parts of autoscaler into there.
  * I'd made cluster-autoscaler to recreate the whole `Autoscaler` instance after each change in the dynamic config read from a configmap.
  * Moved `scale_down*.go`, `scale_up*.go` and `utils*go` into the `autoscaler` package accordingly because they seemed to be meant to be collocated in the same package as the core of CA (which is now implemented as `Autoscaler`)
* `DynamicAutoscaler` is a wrapper around `Autoscaler` which achieves reconfiguration of CA by recreating an `Autoscaler` instance on a configmap change.
* Moved the `createEventRecorder` func from the `main` package to the `utils/kubernetes` package to make it importable from both `main` and `autoscaler`

**Comment by mumoshu** [comment]
sorry to rush you but ping @mwielgus @andrewsykim

**Comment by mwielgus** [comment]
I'm OK with the changes.  I would only rename "autoscaler" to "core" or something like that - i don't like to have two "autoscalers" in the path, one after another. 

Please wait couple more days before resolving the merge conflicts and updating the main loop. We are still modifying the core code and I would like this change to be finished so that nothing is lost in the refactoring by accident.

**Comment by mumoshu** [comment]
@mwielgus Thanks for your review! Is this PR ok to be rebased, probably after #2325 gets merged?

**Comment by mumoshu** [comment]
The former `autoscaler` package is now named `core`.

**Comment by mwielgus** [comment]
Please squash the commits to just 1.

**Comment by mwielgus** [comment]
And sorry for the need to rebase/resolve conflicts - I refreshed the dependencies and lots of things changed in kubernetes codebase.

**Comment by mumoshu** [comment]
@mwielgus I have managed to rebase/resolve conflicts and made a tiny refactoring to ease unit-testing in the future. I had tried to write a bunch of unit-tests but I couldn't finish those in a timely manner, unfortunately.
What we have for now is only `dynamic_autoscaler_test.go`

One of those tests named `autoscaler_test.go` which I had given up looked like:

```golang
package core

import (
	"testing"

	"k8s.io/contrib/cluster-autoscaler/config/dynamic"

	"k8s.io/kubernetes/pkg/client/record"
	"k8s.io/kubernetes/pkg/client/restclient"
	v1core "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/core/v1"
	v1corefake "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/core/v1/fake"
	v1beta1extensions "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/extensions/v1beta1"
	v1beta1extensionsfake "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/extensions/v1beta1/fake"

	"github.com/stretchr/testify/assert"
	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset/fake"
)

type TestClientset struct {
	*fake.Clientset
}

// Core is required to return a fake CoreV1 object with customized RESTClient() func
func (c *TestClientset) Core() v1core.CoreV1Interface {
	return &testCoreV1{FakeCoreV1: &(v1corefake.FakeCoreV1{Fake: &c.Fake})}
}

// Extensions is required to return a fake CoreV1 object with customized RESTClient() func
func (c *TestClientset) Extensions() v1beta1extensions.ExtensionsV1beta1Interface {
	return &testExtensionsV1beta1{FakeExtensionsV1beta1: &(v1beta1extensionsfake.FakeExtensionsV1beta1{Fake: &c.Fake})}
}

type testCoreV1 struct {
	*v1corefake.FakeCoreV1
}

// RESTClient is required to avoid panic when running `cache.NewListWatchFromClient(kubeClient.Core().RESTClient(), ...)
// See `listers.go` to see relevant block which involves this snippet of code
func (c *testCoreV1) RESTClient() restclient.Interface {
	client := &(restclient.RESTClient{})
	return client
}
type testExtensionsV1beta1 struct {
	*v1beta1extensionsfake.FakeExtensionsV1beta1
}

// RESTClient is required to avoid panic when running `cache.NewListWatchFromClient(factory.client.Extensions().RESTClient(), ...)`
// See `listers.go` to see relevant block which involves this snippet of code
func (c *testExtensionsV1beta1) RESTClient() restclient.Interface {
	client := &(restclient.RESTClient{})
	return client
}

func TestNewAutoscalerStatic(t *testing.T) {
	kubeClient := &TestClientset{Clientset: &fake.Clientset{}}
	kubeEventRecorder := record.NewFakeRecorder(0)
	opts := AutoscalerOptions{
		ConfigFetcherOptions: dynamic.ConfigFetcherOptions{
			ConfigMapName: "",
		},
	}
	a := NewAutoscaler(opts, kubeClient, kubeEventRecorder)
	assert.IsType(t, &StaticAutoscaler{}, a)
}

func TestNewAutoscalerDynamic(t *testing.T) {
	kubeClient := &TestClientset{Clientset: &fake.Clientset{}}
	kubeEventRecorder := record.NewFakeRecorder(0)
	opts := AutoscalerOptions{
		ConfigFetcherOptions: dynamic.ConfigFetcherOptions{
			ConfigMapName: "testconfigmap",
		},
	}
	a := NewAutoscaler(opts, kubeClient, kubeEventRecorder)
	assert.IsType(t, &DynamicAutoscaler{}, a)
}
```

in which I couldn't solve panics originating from nil references from somewhere 😢

File: cluster-autoscaler/main.go [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+	kube_flag.InitFlags()
+
+	glog.Infof("Cluster Autoscaler %s", ClusterAutoscalerVersion)
+
+	correctEstimator := false
+	for _, availableEstimator := range estimator.AvailableEstimators {
+		if *estimatorFlag == availableEstimator {
+			correctEstimator = true
+		}
+	}
```
Comment (mumoshu) [inline cluster-autoscaler/main.go]: I tempted to move this validation logic of the estimatorFlag into `estimator` package, probably as something like `func IsValidName(estimatorName string) bool`, but kept untouched thinking doing so would be out of scope of this pull request.

File: cluster-autoscaler/core/dynamic_autoscaler.go [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+// DynamicAutoscaler is a variant of autoscaler which supports dynamic reconfiguration at runtime
+type DynamicAutoscaler struct {
+	autoscaler Autoscaler
+	autoscalerBuilder    AutoscalerBuilder
+	configFetcher dynamic.ConfigFetcher
+}
+
+// NewDynamicAutoscaler builds a DynamicAutoscaler from required parameters
+func NewDynamicAutoscaler(autoscalerBuilder AutoscalerBuilder, configFetcher dynamic.ConfigFetcher) *DynamicAutoscaler {
```
Comment (mumoshu) [inline cluster-autoscaler/core/dynamic_autoscaler.go]: I began to think that we can completely hide `DynamicAutoscaler`, `NewDynamicAutoscaler`, `StaticAutoscaler` and `NewStaticAutoscaler` from other packages by making these private i.e. `dynamicAutoscaler` and `newDynamicAutoscaler` respectively, so that we can ensure that `NewAutoscaler` is used in `main.go`.

File: cluster-autoscaler/utils/kubernetes/factory.go [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+	v1core "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/core/v1"
+
+	"github.com/golang/glog"
+)
+
+// CreateEventRecorder creates an event recorder to send custom events to Kubernetes to be recorded for targeted Kubernetes objects
+func CreateEventRecorder(kubeClient kube_client.Interface) kube_record.EventRecorder {
+	eventBroadcaster := kube_record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.Core().Events("")})
```
Comment (mumoshu) [inline cluster-autoscaler/utils/kubernetes/factory.go]: I couldn't manage to run through this func with a fake kube client.
After digging a bit, I believe that this line relates to the source of problem observed in https://github.com/kubernetes/contrib/pull/2226#issuecomment-274432557

When `kubeClient` is a fake client, `StartRecordingToSink` ends up calling [`func (c *FakeCoreV1) RESTClient() `](https://github.com/kubernetes/kubernetes/blob/master/federation/client/clientset_generated/federation_internalclientset/typed/core/internalversion/fake/fake_core_client.go#L49-L54) which returns nil and then eventually cause a panic from nil dereference errors.

**Comment by mumoshu** [comment]
@mwielgus Added several comments and implementation notes.
Would you mind reviewing again? :bow:

**Comment by mumoshu** [comment]
Updated the code to fix the build errors:

```
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/config/dynamic/config_fetcher.go:31:1: comment on exported function NewConfigFetcher should be of the form "NewConfigFetcher ..."
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/config/dynamic/node_group_spec.go:9:1: comment on exported type NodeGroupSpec should be of the form "NodeGroupSpec ..." (with optional leading article)
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/config/dynamic/node_group_spec.go:19:1: comment on exported function SpecFromString should be of the form "SpecFromString ..."
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/config/dynamic/node_group_spec.go:49:1: comment on exported method NodeGroupSpec.Validate should be of the form "Validate ..." /home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/core/dynamic_autoscaler.go:62:6: exported type AutoscalerBuilder should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/core/dynamic_autoscaler.go:76:1: comment on exported function NewAutoscalerBuilder should be of the form "NewAutoscalerBuilder ..."
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/core/static_autoscaler.go:187:10: if block ends with a return statement, so drop this else and outdent its block /home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/expander/factory/expander_factory.go:10:1: exported function ExpanderStrategyFromString should have comment or be unexported /home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/metrics/metrics.go:61:1: exported function UpdateDuration should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/metrics/metrics.go:66:1: exported function UpdateLastTime should have comment or be unexported
```

**Comment by mumoshu** [comment]
Hi @mwielgus, is there anything I can do to get this reviewed? :bow:

File: cluster-autoscaler/core/dynamic_autoscaler.go (hunk 18 above)
Comment (mwielgus) [inline cluster-autoscaler/core/dynamic_autoscaler.go]: I don't think it is needed right now.

File: cluster-autoscaler/main.go (hunk 17 above)
Comment (mwielgus) [inline cluster-autoscaler/main.go]: Ok. Let's do it in a separate PR.

File: cluster-autoscaler/utils/kubernetes/factory.go (hunk 19 above)
Comment (mwielgus) [inline cluster-autoscaler/utils/kubernetes/factory.go]: We didn't have problems with event recorder before we use them on fake clients in multiple places in the tests, for example here: https://github.com/kubernetes/contrib/blob/master/cluster-autoscaler/scale_down_test.go#L252.

**Comment by mwielgus** [comment]
Reviewed. We are almost there. Sorry for the delay.

File: cluster-autoscaler/core/dynamic_autoscaler.go (hunk 18 above)
Comment (mumoshu) [inline cluster-autoscaler/core/dynamic_autoscaler.go]: OK. Thanks for your confirmation :bow:

File: cluster-autoscaler/main.go (hunk 17 above)
Comment (mumoshu) [inline cluster-autoscaler/main.go]: ok will do!

File: cluster-autoscaler/utils/kubernetes/factory.go (hunk 19 above)
Comment (mumoshu) [inline cluster-autoscaler/utils/kubernetes/factory.go]: Thanks! Turns out that it is caused by the fact I'm blindly creating various podListers and nodeListers like:

```yaml
func NewStaticAutoscaler(opts AutoscalingOptions, predicaterChecker *simulator.PredicateChecker, kubeClient kube_client.Interface, kubeEventRecorder kube_record.EventRecorder) *StaticAutoscaler {
	autoscalingContext := NewAutoscalingContext(opts, predicaterChecker, kubeClient, kubeEventRecorder)

	scaleDown := NewScaleDown(autoscalingContext)
	stopchannel := make(chan struct{})
	unschedulablePodLister := kube_util.NewUnschedulablePodLister(kubeClient, stopchannel)
	scheduledPodLister := kube_util.NewScheduledPodLister(kubeClient, stopchannel)
	readyNodeLister := kube_util.NewReadyNodeLister(kubeClient, stopchannel)
	allNodeLister := kube_util.NewAllNodeLister(kubeClient, stopchannel)
```

which fails when kubeClient is a fake one.

Almost certainly I'd need a way to inject those dependencies via a set of parameters possible wrapped inside a struct.
Do you want me to do that in this PR?

**Comment by mumoshu** [comment]
Hi @mwielgus!
I've managed to finish the fix mentioned in my prev comment https://github.com/kubernetes/contrib/pull/2226#discussion_r98849582.
 I'm not yet satisfied with the verbosity around factory methods like `NewAutoscaler` and test setups in `autoscaler_test.go` but the core of CA is now testable i.e. loosely-coupled enough to be unit-tested, at least.

File: cluster-autoscaler/utils/kubernetes/factory.go (hunk 19 above)
Comment (mumoshu) [inline cluster-autoscaler/utils/kubernetes/factory.go]: fyi my above comment is already addressed in the last commit for this PR

**Comment by mumoshu** [comment]
@mwielgus I've rebased this to the latest master.

**Comment by mumoshu** [comment]
@mwielgus Any chance you could take a look into this again?

**Comment by mwielgus** [comment]
@mumoshu - sorry for the delay. I was on vacation when you addressed the comments. I will take a look tomorrow.

File: cluster-autoscaler/core/static_autoscaler.go [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+		ListerRegistry:           listerRegistry,
+		lastScaleUpTime:          time.Now(),
+		lastScaleDownFailedTrial: time.Now(),
+		scaleDown:                scaleDown,
+	}
+}
+
+// RunOnce iterates over node groups and scales them up/down if necessary
+func (a *StaticAutoscaler) RunOnce(currentTime time.Time) {
+	readyNodeLister := a.readyNodeLister
```
Comment (mwielgus) [inline cluster-autoscaler/core/static_autoscaler.go]: Please sync this with the current main.

File: cluster-autoscaler/core/autoscaling_context.go [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+	Recorder kube_record.EventRecorder
+	// PredicateChecker to check if a pod can fit into a node.
+	PredicateChecker *simulator.PredicateChecker
+	// ExpanderStrategy is the strategy used to choose which node group to expand when scaling up
+	ExpanderStrategy expander.Strategy
+}
+
+// AutoscalingOptions contain various options to customize how autoscaling works
+type AutoscalingOptions struct {
+	// MaxEmptyBulkDelete is a number of empty nodes that can be removed at the same time.
```
Comment (mwielgus) [inline cluster-autoscaler/core/autoscaling_context.go]: Please make sure that the code is fully synced with the current codebase.

Comment (mumoshu) [inline cluster-autoscaler/core/autoscaling_context.go]: Thanks for your comment!
I'll take a look again but could you share me if you already caught any non-syncing parts?

I'd been already resolving conflicts periodically while also surveying recent commit diffs each time after a rebase to keep this PR in sync with master.

**Comment by mwielgus** [comment]
if you synced it recently then i think that the only differences would be minor changes around clusterstate/get candidates for deletion in cluster_autoscaler.go and scale_down.go

I'm sorry it is taking so long. The PR looks good. I can fix some minor issues later.
Please resolve the conflicts and we are good to go. I will not merge anything so that you will not have to resolve yet again. But please do it today.

**Comment by mumoshu** [comment]
Thanks @mwielgus.
I diff'ed the former `cluster_autoscaler.go` vs the current `main.go` and `core/static_autoscaler.go` respectively and caught only typo fixes. I believe that everything is synced now.

**Comment by mwielgus** [comment]
--- FAIL: TestRunOnceWhenNoUpdate (0.00s)
panic: interface conversion: *core.AutoscalerMock is not core.Autoscaler: missing method CleanUp [recovered]
	panic: interface conversion: *core.AutoscalerMock is not core.Autoscaler: missing method CleanUp

**Comment by mumoshu** [comment]
@mwielgus Sorry, it slipped in by mistake. The test is now fixed.

**Comment by mwielgus** [comment]
LGTM

**Comment by mwielgus** [comment]
@mumoshu Merging ... 
Thank you for this contribution and sorry it took so long.

**Comment by mwielgus** [comment]
```
goroutine 30 [running]:
panic(0x17b8ae0, 0xc420016040)
        /usr/local/google/home/mwielgus/.gvm/gos/go1.7.1/src/runtime/panic.go:500 +0x1a1
k8s.io/contrib/cluster-autoscaler/utils/kubernetes.(*ReadyNodeLister).List(0x0, 0x0, 0x0, 0x0, 0x101, 0x8bb2c97000)
        /usr/local/google/home/mwielgus/ca-new/src/k8s.io/contrib/cluster-autoscaler/utils/kubernetes/listers.go:15
7 +0x3a
k8s.io/contrib/cluster-autoscaler/core.(*StaticAutoscaler).CleanUp(0xc420150700)
        /usr/local/google/home/mwielgus/ca-new/src/k8s.io/contrib/cluster-autoscaler/core/static_autoscaler.go:65 +
0x33
main.run(0xc4203507e0)
        /usr/local/google/home/mwielgus/ca-new/src/k8s.io/contrib/cluster-autoscaler/main.go:160 +0x1dc
created by k8s.io/contrib/cluster-autoscaler/vendor/k8s.io/kubernetes/pkg/client/leaderelection.(*LeaderElector).Ru
n
        /usr/local/google/home/mwielgus/ca-new/src/k8s.io/contrib/cluster-autoscaler/vendor/k8s.io/kubernetes/pkg/c
lient/leaderelection/leaderelection.go:150 +0x97
```

**Comment by mumoshu** [comment]
@mwielgus @MaciekPytel Thanks for various follow-ups after merging this.
I'm sorry for breaking things badly but I hope it helped to move things forward at least.
Let me also add that I'm available to be assigned future bug fixes related to this PR.
